pkg/accesscontrol/policy_rule_index.go

```diff
@@ -85,25 +87,29 @@ func (p *policyRuleIndex) get(subjectName string) *AccessSet {
 
 func (p *policyRuleIndex) addAccess(accessSet *AccessSet, namespace string, roleRef rbacv1.RoleRef) {
 	for _, rule := range p.getRules(namespace, roleRef) {
-		for _, group := range rule.APIGroups {
-			for _, resource := range rule.Resources {
-				names := rule.ResourceNames
-				if len(names) == 0 {
-					names = []string{All}
-				}
-				for _, resourceName := range names {
-					for _, verb := range rule.Verbs {
-						accessSet.Add(verb,
-							schema.GroupResource{
-								Group:    group,
-								Resource: resource,
-							}, Access{
-								Namespace:    namespace,
-								ResourceName: resourceName,
-							})
+		if len(rule.Resources) > 0 {
+			for _, group := range rule.APIGroups {
```

Comment:
Could we perhaps extract some of this to their own methods so we reduce the nesting a bit?

Reply:
I'm sorry, what do you mean exactly ?

Reply:
Basically, we're doing nested-ifs and nested-loops 6 levels deep which affects the readability. My suggestion would be to split it a little in two functions to make it more readable. eg: the nonResourceURL case and the resourceURL case.

Reply:
Created another function just to handle resource based rules. It became this way:

https://github.com/rancher/steve/blob/026f994f5803054edb249ac5f2080457b4b50c18/pkg/accesscontrol/policy_rule_index.go#L89-L94